drv/lpc55-sprot-server/src/main.rs

```diff
@@ -175,280 +111,251 @@ fn main() -> ! {
+                        // a major problem. Either we somehow got desynchronized
+                        // or the SP is confused about our capabilities. If there
+                        // is an overrun, we're likely desynchronized. However, an
+                        // overrun should give us fewer bytes to receive, not more
+                        // so the SP must be confused about our buffer size unless
+                        // it's already asserted CSn for the next message.
+                        self.stats.rx_protocol_error_too_many_bytes = self
+                            .stats
+                            .rx_protocol_error_too_many_bytes
+                            .wrapping_add(1);
+                        err = Some(IoError::ProtocolError);
                     }
                 }
-                if !inframe {
-                    // If CSn was deasserted, then the IO loop that was just
-                    // completed would have fetched the remaining bytes out of
-                    // FIFORD and our work is done.
-                    //
-                    // The SP is allowed to send a long message to us at the
-                    // same time it is retrieving a short response from us.
-                    //
-                    // We keep FIFOWR full with zeros when we run out of Tx data
-                    // in order to avoid Tx underrun errors.
-                    //
-                    // So, there will always be "unsent" bytes in FIFOWR.
-                    //
-                    // Actual transmitted bytes are always going to equal the
-                    // number of received bytes. Since the rx_count is not
-                    // skewed by the Tx trailing bytes still in FIFOWR, just
-                    // use rx_count for both.
-
-                    // Update Tx state and account for trailing bytes
-                    // left in FIFOWR if present.
-                    //
-                    // If we cared about those remaining FIFOWR bytes:
-                    // let txremainder = fifostat.txlvl().bits() as usize;
-                    break 'outer;
+            }
+
+            // Let's check for any problems
+            let fifostat = self.spi.fifostat();
+
+            if fifostat.txerr().bit() {
+                // We don't do anything with tx errors other than record them
+                // The SP will see a checksum error if this is a reply, or the
+                // underrun happened after the number of reply bytes and it
+                // doesn't matter.
+                self.spi.txerr_clear();
+                self.stats.tx_underrun = self.stats.tx_underrun.wrapping_add(1);
+            }
+
+            if fifostat.rxerr().bit() {
+                // Rx errors are more important. They mean we're missing
+                // data. We should report this to the SP. This can be used to
+                // potentially throttle sends in the future.
```

Comment:
I'm trying to figure out how to make this actionable. One option is that I can pass the error into the message handler, and if the message doesn't parse correctly, instead of returning an error like `InvalidCrc`, I'd return `FlowError`. A similar thing could be done for when too many bytes are received from the SP.

Alternatively, we can just remove those two error cases altogether and never send "FlowError" or "ProtocolError" and increment the stats counters. I think I prefer the latter, and that also removes 2 of the already large error returns.

Reply:
I could rescind my comment here, either way is valid. Even if the flow error is past the valid message in the Rx buffer, it means that the code didn't keep up and we should make improvements.

I'm thinking in terms of future DMA, which barring quirks, will do all the IO and then we discover how it went.

I like seeing the error codes on the logic analyzer. That would help us search for timing patterns where there were failures. All we really need to know is if the SP should retry or not but I like knowing why.

Reply:
I've been thinking more about this. I can certainly add code to call the handler and return `FlowError` in the case the handler returns an error, but I think this will always occur unless the SPs behavior is flat out wrong. In other words, I'm having a hard time seeing how we'd get a correct message and *then* get an overflow after that. I think this can only happen with a second message getting clocked in while we we are in our tight loop. The SP should not be sending messages back to back, although, sure I suppose it could. The important point is that if we make the change you are suggesting here, we will actually return a successful reply for the first message. Then we will go to handle the next message. It's possible all the bytes were clocked in in the prior tight loop along with the prior overflow. Then we will immediately see CSn deasserted, call the handler, and get a **different** error from `FlowError`. So by not reporting the `FlowError` on the first message we will swallow it and never see it on the logic analyzer. Because of this, I think it's better to just leave things as they are, which is also less code.